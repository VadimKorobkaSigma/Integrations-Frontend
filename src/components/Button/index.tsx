import * as React from 'react';
import cn from 'classnames';

import styles from './styles.module.scss';

interface Props {
    children: string | JSX.Element | JSX.Element[];
    className?: string;
    type?: 'submit';
    onClick?: React.DOMAttributes<HTMLButtonElement>['onClick'];
}

const Button: React.FC<Props> = ({ children, ...args }) => {
    return (
        <button {...{ ...args }} className={cn(styles.button, args.className)}>
            {children}
        </button>
    );
};

export default React.memo(Button);
