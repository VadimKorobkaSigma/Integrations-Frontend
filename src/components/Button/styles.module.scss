.button {
  color: #fff;
  background-color: #2991f3;
  margin-bottom: 1rem;
  width: 160px;
  border: 0;
  padding: 0.5rem 1rem;
  box-shadow: 0px 2px 4px rgba(0, 0, 0, 0.15);
  border-radius: 4px;
  display: flex;
  align-items: center;
  justify-content: flex-start;
  font-size: 20px;

  &:hover {
    background-color: #1579d7;
  }
}
