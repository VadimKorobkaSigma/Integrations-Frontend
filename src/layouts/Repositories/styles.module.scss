.column {
  display: flex;
  padding: 20px;
  flex-direction: column;
  overflow: hidden;

  h3 {
    line-height: 39px;
    color: #565360;
  }

  ul {
    overflow: auto;
  }
}

.item {
  display: flex;
  justify-content: space-between;
  margin-bottom: 15px;

  .name {
    display: flex;
    align-items: center;

    span {
      text-overflow: ellipsis;
      overflow: hidden;
      white-space: nowrap;
      font-size: 18px;
      line-height: 24px;
      letter-spacing: -0.005em;
      color: #565360;

      &.active {
        border-bottom: 1px solid;
        font-weight: 500;
      }
    }
  }

  .webhook {
    width: initial;
    margin-bottom: 0;
  }

  &:not(:last-child) {
    border-bottom: 1px solid #f6f6f6;
    padding-bottom: 10px;
  }
}
