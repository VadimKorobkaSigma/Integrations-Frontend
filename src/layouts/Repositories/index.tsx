import * as React from 'react';
import cn from 'classnames';
import SVG from 'react-inlinesvg';

import { Repository } from '@dtos/repository';
import api from '@services/api';
import Button from '@components/Button';

import trashIcon from '@assets/images/trash.svg';
import webhookIcon from '@assets/images/webhook.svg';
import orgIcon from '@assets/images/orgIcon.svg';
import styles from './styles.module.scss';
import useError from '@services/hooks/useError';
import ErrorComponent from '@components/ErrorComponent';

interface OwnProps {
    selectedOrg: string;
}

type Props = OwnProps;

const Repositories: React.FC<Props> = ({ selectedOrg }) => {
    const [error, handleError, clearError] = useError();
    const [isLoading, setIsLoading] = React.useState(false);
    const [repos, setRepos] = React.useState<Repository[]>([]);

    React.useEffect(() => {
        let isCancelled = false;
        if (selectedOrg) {
            setIsLoading(true);
            api.getOrganizationRepos(selectedOrg)
                .then((data) => {
                    if (!isCancelled) {
                        setRepos(data);
                        clearError();
                    }
                })
                .catch(handleError)
                .finally(() => setIsLoading(false));
        }

        return () => {
            isCancelled = true;
        };
    }, [selectedOrg]);

    const toggleWebhook = async (rep: Repository) => {
        try {
            let newWebhookId = '';
            if (rep.webhookEnabled) {
                await api.removeWebhook(selectedOrg, rep.id, rep.webhookId);
            } else {
                newWebhookId = await api.installWebhook(selectedOrg, rep.id).then(({ id }) => id);
            }

            setRepos(
                repos.map((repository) =>
                    repository.id === rep.id
                        ? { ...rep, webhookEnabled: !rep.webhookEnabled, webhookId: newWebhookId }
                        : repository,
                ),
            );
            clearError();
        } catch (err) {
            handleError(err);
        }
    };

    return (
        <section className={styles.column}>
            <h2 className={styles.header}>Repositories</h2>
            <ErrorComponent error={error} />
            {isLoading && <h3 className={styles.message}> Loading...</h3>}
            {!isLoading && !selectedOrg && <h3 className={styles.message}>Please select organization.</h3>}
            {!isLoading && selectedOrg && !repos.length ? (
                <h3 className={styles.message}> No repositories available.</h3>
            ) : null}
            {!isLoading && repos.length ? (
                <ul>
                    {repos.map((rep) => (
                        <li key={rep.id} className={styles.item}>
                            <div className={styles.name}>
                                <img src={orgIcon} alt="orgIcon" width={24} />
                                <span>{rep.name}</span>
                            </div>
                            <Button
                                className={cn(styles.webhook, rep.webhookEnabled && styles.remove)}
                                onClick={() => toggleWebhook(rep)}
                            >
                                <SVG src={rep.webhookEnabled ? trashIcon : webhookIcon} />
                                <p>{rep.webhookEnabled ? 'Stop Cx Scan' : 'Start Cx Scan'}</p>
                            </Button>
                        </li>
                    ))}
                </ul>
            ) : null}
        </section>
    );
};

export default React.memo(Repositories);
