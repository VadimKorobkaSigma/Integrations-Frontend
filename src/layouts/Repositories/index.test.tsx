import * as React from 'react';
import Repositories from './index';
import { shallow } from 'enzyme';
import Button from '@components/Button';
const data = [
    { id: 'test_id', name: 'test 1', webhookEnabled: false },
    { id: 'test_id_2', name: 'test 2', webhookEnabled: true, webhookId: 1 },
];
test('Repositories Component', () => {
    global.fetch = jest.fn().mockImplementation(() =>
        Promise.resolve({
            json: () => Promise.resolve(data),
        }),
    );
    jest.spyOn(React, 'useState').mockImplementation(
        jest
            .fn()
            .mockReturnValueOnce(['', jest.fn()])
            .mockReturnValueOnce([false, jest.fn()])
            .mockReturnValueOnce([data, jest.fn()])
            .mockReturnValueOnce(['', jest.fn()])
            .mockReturnValueOnce([data, jest.fn()])
            .mockReturnValue([null, jest.fn()]),
    );
    const wrapper = shallow(<Repositories selectedOrg="test_id" />);
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.find(Button).first().simulate('click');
});
