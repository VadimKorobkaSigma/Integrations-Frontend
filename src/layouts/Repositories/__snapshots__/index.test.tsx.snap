// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repositories Component 1`] = `"<section><h2>Repositories</h2><div class=\\"\\"><input placeholder=\\"Search\\" value=\\"\\" class=\\"\\"/></div><ul><li><div><img alt=\\"orgIcon\\" width=\\"24\\"/><span><span>test 1</span></span></div><button class=\\"\\"><p>Setup project</p></button></li><li><div><img alt=\\"orgIcon\\" width=\\"24\\"/><span><span>test 2</span></span></div><button class=\\"\\"><p>Remove setup</p></button></li></ul></section>"`;
